File: app/src/main/java/com/github/quentin7b/sn/ui/view/StickyNoteFullView.java
package com.github.quentin7b.sn.ui.view;

import android.animation.Animator;
import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Build;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.AppCompatCheckBox;
import android.support.v7.widget.AppCompatTextView;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.widget.LinearLayout;

import com.github.quentin7b.sn.ColorHelper;
import com.github.quentin7b.sn.R;
import com.github.quentin7b.sn.Tool;
import com.github.quentin7b.sn.database.model.StickyNotification;
import com.wdullaer.materialdatetimepicker.date.DatePickerDialog;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class StickyNoteFullView extends LinearLayout {

    Integer colorRes;
    @Nullable
    @BindView(R.id.note_title_et)
    TextInputEditText noteTitleEt;
    @BindView(R.id.notification_cb)
    AppCompatCheckBox showNotificationCb;
    @BindView(R.id.level_btn)
    LabelImageButton levelImageBtn;
    @BindView(R.id.note_content_et)
    TextInputEditText noteContentEt;
    @BindView(R.id.note_content_tv)
    AppCompatTextView dateTextView;

    private SimpleDateFormat dateFormat;


    public StickyNoteFullView(Context context) {
        super(context);
        initLayout();
    }

    public StickyNoteFullView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initLayout();
    }

    public StickyNoteFullView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initLayout();
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public StickyNoteFullView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        initLayout();
    }

    private void initLayout() {
        Context context = getContext();
        dateFormat = new SimpleDateFormat(context.getString(R.string.long_date_format),
                Tool.getLocale(context));

        setOrientation(LinearLayout.VERTICAL);
        LayoutInflater.from(getContext()).inflate(R.layout.view_note_full, this, true);
        ButterKnife.bind(this);
        setSaveEnabled(true);
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        Bundle instanceState = new Bundle();
        instanceState.putParcelable(EXTRA.SUPER, super.onSaveInstanceState());
        instanceState.putString(EXTRA.TITLE, getTitle());
        instanceState.putString(EXTRA.CONTENT, getContent());
        instanceState.putBoolean(EXTRA.NOTIFICATION, isNotification());
        instanceState.putInt(EXTRA.DEFCON, getDefcon().describe());
        instanceState.putLong(EXTRA.DATE, getDate() != null ? getDate().getTime() : -1);
        return super.onSaveInstanceState();
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        if (state instanceof Bundle) {
            Bundle bundle = (Bundle) state;
            setTitle(bundle.getString(EXTRA.TITLE, ""));
            setContent(bundle.getString(EXTRA.CONTENT, ""));
            setNotification(bundle.getBoolean(EXTRA.NOTIFICATION, true));
            setDefcon(StickyNotification.Defcon.from(bundle.getInt(EXTRA.DEFCON, StickyNotification.Defcon.NORMAL.describe())));
            state = bundle.getParcelable(EXTRA.SUPER);
            long dateLong = bundle.getLong(EXTRA.DATE);
            if (dateLong != -1) {
                setDate(new Date(dateLong));
            }
        }
        super.onRestoreInstanceState(state);
    }

    @OnClick(R.id.level_btn)
    void onDefconClick() {
        final View dialogView = LayoutInflater.from(getContext()).inflate(R.layout.dialog_color_picker, null);
        final AlertDialog alertDialog = new AlertDialog.Builder(getContext())
                .setView(dialogView)
                .create();

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
            alertDialog.setOnShowListener(new DialogInterface.OnShowListener() {
                @SuppressLint("NewApi")
                @Override
                public void onShow(final DialogInterface dialog) {
                    final View view = dialogView.findViewById(R.id.reveal_view);
                    int w = view.getWidth();
                    int h = view.getHeight();
                    float maxRadius = (float) Math.sqrt(w * w / 2 + h * h / 2);
                    Animator revealAnimator = ViewAnimationUtils.createCircularReveal(view,
                            w, h / 2, 0, maxRadius);
                    view.setVisibility(View.VISIBLE);
                    revealAnimator.start();
                }
            });
        }

        LabelImageButton uselessBtn = (LabelImageButton) dialogView.findViewById(R.id.useless_btn);
        LabelImageButton normalBtn = (LabelImageButton) dialogView.findViewById(R.id.normal_btn);
        LabelImageButton importantBtn = (LabelImageButton) dialogView.findViewById(R.id.important_btn);
        LabelImageButton ultraBtn = (LabelImageButton) dialogView.findViewById(R.id.ultra_btn);

        uselessBtn.setColorRes(ColorHelper.getDefconColor(StickyNotification.Defcon.USELESS));
        normalBtn.setColorRes(ColorHelper.getDefconColor(StickyNotification.Defcon.NORMAL));
        importantBtn.setColorRes(ColorHelper.getDefconColor(StickyNotification.Defcon.IMPORTANT));
        ultraBtn.setColorRes(ColorHelper.getDefconColor(StickyNotification.Defcon.ULTRA));

        View.OnClickListener colorClickListener = new View.OnClickListener() {

            @Override
            public void onClick(View v) {
                LabelImageButton imageButton = (LabelImageButton) v;
                StickyNoteFullView.this.colorRes = imageButton.getColorRes();
                alertDialog.dismiss();
            }
        };

        uselessBtn.setOnClickListener(colorClickListener);
        normalBtn.setOnClickListener(colorClickListener);
        importantBtn.setOnClickListener(colorClickListener);
        ultraBtn.setOnClickListener(colorClickListener);

        alertDialog.setOnDismissListener(new DialogInterface.OnDismissListener() {
            @Override
            public void onDismiss(DialogInterface dialog) {
                StickyNoteFullView.this.levelImageBtn.setColorRes(StickyNoteFullView.this.colorRes);
            }
        });

        alertDialog.setCancelable(true);
        alertDialog.show();
    }

    @OnClick(R.id.note_content_tv)
    void onDateClick() {
        Calendar calendar = Calendar.getInstance();
        Date currentDate = getDate();
        if (currentDate != null) {
            calendar.setTime(currentDate);
        }
        DatePickerDialog dpd = DatePickerDialog.newInstance(
                new DatePickerDialog.OnDateSetListener() {
                    @Override
                    public void onDateSet(DatePickerDialog view, int year, int monthOfYear, int dayOfMonth) {
                        Calendar setCalendar = Calendar.getInstance();
                        setCalendar.set(Calendar.YEAR, year);
                        setCalendar.set(Calendar.MONTH, monthOfYear);
                        setCalendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                        setDate(setCalendar.getTime());
                    }
                },
                calendar.get(Calendar.YEAR),
                calendar.get(Calendar.MONTH),
                calendar.get(Calendar.DAY_OF_MONTH)
        );
        dpd.show(((Activity) getContext()).getFragmentManager(), "Datepickerdialog");
    }

    public String getTitle() {
        if (noteTitleEt != null) {
            return noteTitleEt.getText().toString();
        } else {
            return null;
        }
    }

    public void setTitle(String title) {
        if (noteTitleEt != null) {
            noteTitleEt.setText(title);
            noteTitleEt.setSelection(title.length());
        }
    }

    public StickyNotification.Defcon getDefcon() {
        return ColorHelper.getColorDefcon(colorRes);
    }

    public void setDefcon(StickyNotification.Defcon defcon) {
        colorRes = ColorHelper.getDefconColor(defcon);
        levelImageBtn.setColorRes(colorRes);
    }

    public String getContent() {
        return noteContentEt.getText().toString();
    }

    public void setContent(String content) {
        noteContentEt.setText(content);
        noteContentEt.setSelection(content.length());
    }

    public boolean isNotification() {
        return showNotificationCb.isChecked();
    }

    public void setNotification(boolean isNotification) {
        showNotificationCb.setChecked(isNotification);
    }

    public Date getDate() {
        if (!TextUtils.isEmpty(dateTextView.getText())) {
            try {
                return dateFormat.parse(dateTextView.getText().toString());
            } catch (ParseException e) {
                return null;
            }
        } else {
            return null;
        }
    }

    public void setDate(Date date) {
        if (date != null) {
            dateTextView.setText(dateFormat.format(date));
        }
    }

    private static final class EXTRA {
        private static final String SUPER = "superState";
        private static final String TITLE = "titleState";
        private static final String CONTENT = "contentState";
        private static final String DEFCON = "defconState";
        private static final String NOTIFICATION = "notificationState";
        private static final String DATE = "dateState";
    }

}

File: app/src/main/java/com/github/quentin7b/sn/ui/view/StickyNoteFullView.kt
package com.github.quentin7b.sn.ui.view

import android.annotation.TargetApi
import android.app.Activity
import android.content.Context
import android.os.Build
import android.os.Bundle
import android.os.Parcelable
import android.support.design.widget.TextInputEditText
import android.support.v7.app.AlertDialog
import android.support.v7.widget.AppCompatCheckBox
import android.support.v7.widget.AppCompatTextView
import android.text.TextUtils
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewAnimationUtils
import android.widget.LinearLayout

import com.github.quentin7b.sn.ColorHelper
import com.github.quentin7b.sn.R
import com.github.quentin7b.sn.Tool
import com.github.quentin7b.sn.database.model.StickyNotification
import com.wdullaer.materialdatetimepicker.date.DatePickerDialog
import kotlinx.android.synthetic.main.view_note_full.view.*

import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date


class StickyNoteFullView : LinearLayout {

    internal var colorRes: Int? = null

    private var dateFormat: SimpleDateFormat? = null

    var title: String?
        get() = if (note_title_et != null) {
            note_title_et!!.text.toString()
        } else {
            null
        }
        set(title) {
            if (note_title_et != null) {
                note_title_et!!.setText(title)
                note_title_et!!.setSelection(title!!.length)
            }
        }

    var defcon: StickyNotification.Defcon?
        get() = ColorHelper.getColorDefcon(colorRes)
        set(defcon) {
            colorRes = ColorHelper.getDefconColor(defcon!!)
            level_btn!!.setColorRes(colorRes!!)
        }

    var content: String
        get() = note_content_et!!.text.toString()
        set(content) {
            note_content_et!!.setText(content)
            note_content_et!!.setSelection(content.length)
        }

    var isNotification: Boolean
        get() = notification_cb!!.isChecked
        set(isNotification) {
            notification_cb!!.isChecked = isNotification
        }

    var date: Date?
        get() = if (!TextUtils.isEmpty(note_content_tv!!.text)) {
            try {
                dateFormat!!.parse(note_content_tv!!.text.toString())
            } catch (e: ParseException) {
                null
            }

        } else {
            null
        }
        set(date) {
            if (date != null) {
                note_content_tv!!.text = dateFormat!!.format(date)
            }
        }


    constructor(context: Context) : super(context) {
        initLayout()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        initLayout()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        initLayout()
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        initLayout()
    }

    private fun initLayout() {
        val context = context
        dateFormat = SimpleDateFormat(context.getString(R.string.long_date_format),
                Tool.getLocale(context))

        orientation = LinearLayout.VERTICAL
        LayoutInflater.from(getContext()).inflate(R.layout.view_note_full, this, true)
        isSaveEnabled = true

        level_btn?.setOnClickListener { onDefconClick() }
        note_content_tv?.setOnClickListener { onDateClick() }
    }

    override fun onSaveInstanceState(): Parcelable? {
        val instanceState = Bundle()
        instanceState.putParcelable(EXTRA.SUPER, super.onSaveInstanceState())
        instanceState.putString(EXTRA.TITLE, title)
        instanceState.putString(EXTRA.CONTENT, content)
        instanceState.putBoolean(EXTRA.NOTIFICATION, isNotification)
        instanceState.putInt(EXTRA.DEFCON, defcon!!.describe())
        instanceState.putLong(EXTRA.DATE, if (date != null) date!!.time else -1)
        return super.onSaveInstanceState()
    }

    override fun onRestoreInstanceState(state: Parcelable?) {
        var state = state
        if (state is Bundle) {
            val bundle = state as Bundle?
            title = bundle!!.getString(EXTRA.TITLE, "")
            content = bundle.getString(EXTRA.CONTENT, "")
            isNotification = bundle.getBoolean(EXTRA.NOTIFICATION, true)
            defcon = StickyNotification.Defcon.from(bundle.getInt(EXTRA.DEFCON, StickyNotification.Defcon.NORMAL.describe()))
            state = bundle.getParcelable(EXTRA.SUPER)
            val dateLong = bundle.getLong(EXTRA.DATE)
            if (dateLong != -1L) {
                date = Date(dateLong)
            }
        }
        super.onRestoreInstanceState(state)
    }

    internal fun onDefconClick() {
        val dialogView = LayoutInflater.from(context).inflate(R.layout.dialog_color_picker, null)
        val alertDialog = AlertDialog.Builder(context)
                .setView(dialogView)
                .create()

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
            alertDialog.setOnShowListener {
                val view = dialogView.findViewById<View>(R.id.reveal_view)
                val w = view.width
                val h = view.height
                val maxRadius = Math.sqrt((w * w / 2 + h * h / 2).toDouble()).toFloat()
                val revealAnimator = ViewAnimationUtils.createCircularReveal(view,
                        w, h / 2, 0f, maxRadius)
                view.visibility = View.VISIBLE
                revealAnimator.start()
            }
        }

        val uselessBtn = dialogView.findViewById<View>(R.id.useless_btn) as LabelImageButton
        val normalBtn = dialogView.findViewById<View>(R.id.normal_btn) as LabelImageButton
        val importantBtn = dialogView.findViewById<View>(R.id.important_btn) as LabelImageButton
        val ultraBtn = dialogView.findViewById<View>(R.id.ultra_btn) as LabelImageButton

        uselessBtn.setColorRes(ColorHelper.getDefconColor(StickyNotification.Defcon.USELESS))
        normalBtn.setColorRes(ColorHelper.getDefconColor(StickyNotification.Defcon.NORMAL))
        importantBtn.setColorRes(ColorHelper.getDefconColor(StickyNotification.Defcon.IMPORTANT))
        ultraBtn.setColorRes(ColorHelper.getDefconColor(StickyNotification.Defcon.ULTRA))

        val colorClickListener = OnClickListener { v ->
            val imageButton = v as LabelImageButton
            this@StickyNoteFullView.colorRes = imageButton.getColorRes()
            alertDialog.dismiss()
        }

        uselessBtn.setOnClickListener(colorClickListener)
        normalBtn.setOnClickListener(colorClickListener)
        importantBtn.setOnClickListener(colorClickListener)
        ultraBtn.setOnClickListener(colorClickListener)

        alertDialog.setOnDismissListener { this@StickyNoteFullView.level_btn!!.setColorRes(this@StickyNoteFullView.colorRes!!) }

        alertDialog.setCancelable(true)
        alertDialog.show()
    }

    internal fun onDateClick() {
        val calendar = Calendar.getInstance()
        val currentDate = date
        if (currentDate != null) {
            calendar.time = currentDate
        }
        val dpd = DatePickerDialog.newInstance(
                { view, year, monthOfYear, dayOfMonth ->
                    val setCalendar = Calendar.getInstance()
                    setCalendar.set(Calendar.YEAR, year)
                    setCalendar.set(Calendar.MONTH, monthOfYear)
                    setCalendar.set(Calendar.DAY_OF_MONTH, dayOfMonth)
                    date = setCalendar.time
                },
                calendar.get(Calendar.YEAR),
                calendar.get(Calendar.MONTH),
                calendar.get(Calendar.DAY_OF_MONTH)
        )
        dpd.show((context as Activity).fragmentManager, "Datepickerdialog")
    }

    private object EXTRA {
        val SUPER = "superState"
        val TITLE = "titleState"
        val CONTENT = "contentState"
        val DEFCON = "defconState"
        val NOTIFICATION = "notificationState"
        val DATE = "dateState"
    }

}
